components/num-util/src/fixed_decimal.rs

```diff
@@ -0,0 +1,517 @@
+    ///
+    /// ```
+    /// use icu_num_util::FixedDecimal;
+    ///
+    /// let mut dec = FixedDecimal::from(120);
+    /// assert_eq!(0..=2, dec.magnitude_range());
+    /// ```
+    pub fn magnitude_range(&self) -> RangeInclusive<i16> {
+        RangeInclusive::new(self.lower_magnitude, self.upper_magnitude)
+    }
+
+    /// Shift the digits by a power of 10. Leading or trailing zeros may be added to keep the digit
+    /// at magnitude 0 (the last digit before the decimal separator) visible.
+    ///
+    /// # Example
+    ///
+    /// ```
+    /// use icu_num_util::FixedDecimal;
+    ///
+    /// let mut dec = FixedDecimal::from(42);
+    /// assert_eq!("42", dec.to_string());
+    ///
+    /// dec.adjust_magnitude(3);
+    /// assert_eq!("42000", dec.to_string());
+    /// ```
+    pub fn adjust_magnitude(&mut self, delta: i16) -> Result<(), Error> {
+        if delta > 0 {
+            self.upper_magnitude = match self.upper_magnitude.checked_add(delta) {
+                Some(v) => v,
+                None => return Err(Error::Limit),
+            };
+            // If we get here, then the magnitude change is in-bounds.
+            let lower_magnitude = self.lower_magnitude + delta;
+            self.lower_magnitude = cmp::min(0, lower_magnitude);
+        } else if delta < 0 {
+            self.lower_magnitude = match self.lower_magnitude.checked_add(delta) {
+                Some(v) => v,
+                None => return Err(Error::Limit),
+            };
+            // If we get here, then the magnitude change is in-bounds.
+            let upper_magnitude = self.upper_magnitude + delta;
+            self.upper_magnitude = cmp::max(0, upper_magnitude);
+        }
+        self.magnitude += delta;
+        debug_assert!(self.check_invariants());
+        return Ok(());
+    }
+
+    /// Assert that the invariants among struct fields are enforced. Returns true if all are okay.
+    /// Call this in any method that mutates the struct fields.
+    ///
+    /// Example: `debug_assert!(self.check_invariants())`
+    #[cfg(debug_assertions)]
+    fn check_invariants(&self) -> bool {
+        // magnitude invariants:
+        debug_assert_ge!(self.upper_magnitude, self.magnitude, "{:?}", self);
+        debug_assert_le!(self.lower_magnitude, self.magnitude, "{:?}", self);
+        debug_assert_ge!(self.upper_magnitude, 0, "{:?}", self);
+        debug_assert_le!(self.lower_magnitude, 0, "{:?}", self);
+
+        // digits invariants:
+        let max_len = (self.magnitude as i32 - self.lower_magnitude as i32 + 1) as usize;
+        debug_assert_le!(self.digits.len(), max_len, "{:?}", self);
+        if self.digits.len() > 0 {
+            debug_assert_ne!(self.digits[0], 0, "{:?}", self);
+            debug_assert_ne!(self.digits[self.digits.len() - 1], 0, "{:?}", self);
+        }
+
+        // All OK
+        true
+    }
+}
+
+// TODO(review): This should be fmt::Display, but that is very no_std-unfriendly for code size.
```

Comment:
can you elaborate? Why wouldn't we want `Display` here? If it's not used, it'll be shaken off.

Reply:
Library code might want to use ToString, but if you implement Display, then ToString uses the heavier Display code path.

Reply:
Can you improve the comment then? Maybe:

```
`Display` provides a heavier implementation of `ToString` while we want to explicitly handle `ToString` for library use.
```

Also, can we implement both to allow `FixedDecimal` to be used in alloc-free way?

Reply:
`TODO(review)` means that I wanted you to evaluate a decision I made but that I wasn't 100% confident in.  First, do you agree with me that it is a worthwhile win for code size to separate ToString from Display?  If so, then yes, I can look into also implementing Display.

Reply:
Yep, agree.

Reply:
If I do `impl Display for FixedDecimal`, then the `impl ToString for FixedDecimal` fails to compile with E0119, since the standard library has a generic ToString for types implementing Display.

Reply:
I went ahead and did `impl Display for FixedDecimal`, and supplemented it with new methods `write_to` and `write_len` if one needs a more efficient implementation.